### PR TITLE
Cache arbitrary labels in the SQL database

### PR DESCRIPTION
**Pull request by ericpromislow**
WIP - no unit tests in yet

### CODE REVIEW

File: pkg/cache/sharedinformerfactory.go [hunk 1]
```diff
@@ -125,22 +125,37 @@ func (f *sharedCacheFactory) Start(ctx context.Context) error {
...
+
+	informers := map[schema.GroupVersionKind]cache.SharedIndexInformer{}
+	for informerType, informer := range f.caches {
+		metrics.IncTotalCachedObjects(informerType.Group, informerType.Version, informerType.Kind, float64(len(informer.GetStore().List())))
+		if f.startedCaches[informerType] {
+			informers[informerType] = informer
+		}
+	}
+	return informers
+}
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply pkg/cache/sharedinformerfactory.go:140]: I first read it expecting it to be used as a closure, and it isn't. But sure, I'll restore it.

File: pkg/cache/sql/store/store.go [hunk 2]
```diff
@@ -183,6 +188,63 @@ func (s *Store) Add(obj any) error {
...
+		if strings.HasPrefix(label, prefix) {
+			knownLabelColumns = append(knownLabelColumns, label[prefixLen:len(label)-1])
+		}
+	}
+	newLabels := make([]string, 0, len(incomingLabels))
+	for _, label := range incomingLabels {
+		if !sliceContains(knownLabelColumns, label) {
+			newLabels = append(newLabels, label)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply pkg/cache/sql/store/store.go:218]: I was too, but treat this as a first pass at the problem. I'll look into tracking the labels as a set on each GVK and then don't have to make the sql queries.

File: pkg/cache/sql/informer/indexer.go [hunk 3]
```diff
@@ -135,6 +138,19 @@ func (i *Indexer) AfterUpsert(key string, obj any, tx db.TXClient) error {
...
 	}
 
+	raw, ok := obj.(*unstructured.Unstructured)
+	if ok {
+		labels := raw.GetLabels()
+		keys := make([]string, 0, len(labels))
+		for k := range labels {
+			keys = append(keys, k)
+		}
+		err = i.Store.AddLabels(tx, keys)
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply pkg/cache/sql/informer/indexer.go:148]: Ah, true.  So I need to modify it in the upsert itself...?

**Review by moio — CHANGES_REQUESTED** [review]
Thanks for your contribution @ericpromislow.

I left an inline comment about the general idea behind the implementation of this feature, as I am a bit unsure if this is the best path. Hopefully that conveys the main idea decently, I am happy to discuss it anytime.

**Comment by ericpromislow** [comment]
I finally can get back to this. @moio I agree with you that a separate `node_labels` table sounds better than adding columns to each table, and I have no problem with `join`s.

**Comment by ericpromislow** [comment]
This approach is just wrong. I'm going to create a new PR with a completely different implementation, as suggested here, and will leave this PR as a contrast for reference in the new PR.

**Pull request closed without merging**